### PR TITLE
Improve light level button in the editor

### LINKED ISSUE #3881: Add button to use current light level in the editor and make it the default
add a button for "current" light level (the light level from the stage when the player switched to the editor) as the default view in the editor, and make the patch map use the current editor light level (needs to be applied to all patches for relative comparisons to work) in comparisons so that the patch map and editor visuals are consistent with each other.

### LINKED ISSUE #3919: Editor should remember the selected light level when saving and loading
right now it resets when loading a save, it should be saved like a bunch of editor GUI state is

### PR DESCRIPTION
**Pull request by Kasterisk**
**Brief Description of What This PR Does**

Add "Current" button which syncs the light level in all patch in the patch map to the true daylight level and this is by default.

The editor should also now correctly loads the selected light level button in a save.

Loading old save works, nothing seems broken but more testing pleasee.

**Related Issues**

Fixes #3881.
Fixes #3919.

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/microbe_stage/editor/MicrobeEditor.cs [hunk 1]
```diff
@@ -314,6 +314,12 @@ protected override void SetupEditedSpecies()
         base.SetupEditedSpecies();
     }
 
+    protected override void ApplyComponentLightLevels()
+    {
+        base.ApplyComponentLightLevels();
+        patchMapTab.UpdateShownPatchDetails();
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/MicrobeEditor.cs:320]: Patch map is an editor component? Why doesn't it get given the info through the editor component light level update callback?

Comment (Kasterisk) [reply src/microbe_stage/editor/MicrobeEditor.cs:320]: That didn't cross my mind...

File: src/microbe_stage/editor/CellEditorComponent.cs [hunk 2]
```diff
@@ -1072,8 +1068,9 @@ public float CalculateTotalDigestionSpeed()
 
     public override void OnLightLevelChanged(float lightLevel)
     {
-        var maxLightLevel = Editor.CurrentPatch.GetCompoundAmount("sunlight", CompoundAmountType.Maximum);
-        var templateMaxLightLevel = Editor.CurrentPatch.GetCompoundAmount("sunlight", CompoundAmountType.Template);
+        var maxLightLevel = Editor.CurrentPatch.Biome.Compounds[sunlight].Ambient;
+        var minLightLevel = Editor.CurrentPatch.Biome.MinimumCompounds[sunlight].Ambient;
+        var templateMaxLightLevel = Editor.CurrentPatch.Biome.Compounds[sunlight].Ambient;
```
Comment (hhyyrylainen) [inline src/microbe_stage/editor/CellEditorComponent.cs:1073]: Either max or the template light level variable is wrong here. Or at least that has changed. Is this change intentional?

Comment (Kasterisk) [reply src/microbe_stage/editor/CellEditorComponent.cs:1073]: Yes, the previous method returns percentage, which is not suitable for what we want to do later in the code.

Comment (hhyyrylainen) [reply src/microbe_stage/editor/CellEditorComponent.cs:1073]: Well the variable name has to change then as it is not related to the template. I guess logically the change is fine as well as the player will be more interested in seeing what the light level can be in the patch, not what the archetype biome can have as the light level.

Comment (Kasterisk) [reply src/microbe_stage/editor/CellEditorComponent.cs:1073]: Realized what you're saying was `maxLightLevel` is using `Compounds` instead of `MaximumCompounds`... let me rectify my dumb mistake.

Comment (hhyyrylainen) [reply src/microbe_stage/editor/CellEditorComponent.cs]: Also fix the fact that the "template" max light level doesn't actually have the template light level... The template light level if still actually needed can be retrieved from the Patch but not the biome conditions.

Comment (Kasterisk) [reply src/microbe_stage/editor/CellEditorComponent.cs]: Alright, figured it out, everything should be good.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
The code now looks correct. When playtesting I noticed one bug: when I have a patch selected, go to the cell editor tab, switch the time of day, and finally go back to the patch map tab I see the previous light level values in the right panel. The situation corrects itself when I click on a different patch and then click on the original patch again.

So there's some callback thing missing for updating the currently selected patch data when the light level is changed. That was the only problem I saw, everything else seemed to make sense.

**Comment by Kasterisk** [comment]
Apparently the order of this matter now:
```csharp
    protected override IEnumerable<IEditorComponent> GetAllEditorComponents()
    {
        yield return reportTab;
        yield return patchMapTab;
        yield return cellEditorTab;
    }
```

The patch map tab must follow the cell editor tab so that it gets the correct light level. Should we reorder it or entirely move the patch light level updation out of the cell editor tab?

**Comment by hhyyrylainen** [comment]
Hang on, could you explain the path the update logic takes? Why does the order matter? When the light level change is triggered, that should be sent to all of the components. So it shouldn't matter what order the components are specified in...

**Comment by Kasterisk** [comment]
```csharp
    private void ApplyComponentLightLevels()
    {
        foreach (var editorComponent in GetAllEditorComponents())
        {
            editorComponent.OnLightLevelChanged(lightLevel);
        }
    }
```

Here, `OnLightLevelChanged` is first called in `MicrobeEditorReportComponent`, then in `PatchMapEditorComponent` and finally in `CellEditorComponent`.

`CellEditorComponent` is where the actual light level in all patches is updated, `PatchMapEditorComponent` just updates its details with the new light level. From this, we can deduce that the bug was caused by the light level change callback being triggered in the patch map first instead of the cell editor.

Although reordering `GetAllEditorComponents` might cause a whole lot of other unforeseen problems...

**Comment by hhyyrylainen** [comment]
> CellEditorComponent is where the actual light level in all patches is updated, PatchMapEditorComponent just updates its details with the new light level. From this, we can deduce that the bug was caused by the light level change callback being triggered in the patch map first instead of the cell editor.

I think it's anyway pretty unclean that the cell editor component would update the patch map view's data. Much better for the map to handle that. That's needed anyway for multicellular editor as eventually the player should be able to change the light level in the body plan editor so it is not useful for that map update logic to reside in the cell editor component.

> Although reordering GetAllEditorComponents might cause a whole lot of other unforeseen problems...

Absolutely. My intention is that the order of that list actually matches the tabs in the GUI and logical flow of the user

**Comment by Kasterisk** [comment]
> I think it's anyway pretty unclean that the cell editor component would update the patch map view's data. Much better for the map to handle that. That's needed anyway for multicellular editor as eventually the player should be able to change the light level in the body plan editor so it is not useful for that map update logic to reside in the cell editor component.

Done. That makes total sense.

**Review by hhyyrylainen — APPROVED** [review]
Assuming the one bug I found is now fixed, this is good to go. I can't retest this right now as I'm doing a refactor on another PR.

**Comment by hhyyrylainen** [comment]
I've now re-tested this.
